### ISSUE TITLE
system_auth: use prepared paged statements for reading from system_auth tables

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylla#7415]
*Installation details*
HEAD: 722a5db27e2014d40805dfdaa1035c1018bfb75e

*Description*
Use prepared paged queries for reading from system_auth.
Today when these are read, e.g. during the client (re)connect we see corresponding metrics bumped up and generates unnecessary noise.

**Comment by vladzcloudius** [comment]
@eliransin FYI

**Comment by slivne** [comment]
AFAIK - internal queries are non prepared - we have that true for every where else in the system as well. One of the issues is supporting the case in which they need to  be re-prepaired.  

@kostja / @gleb-cloudius - any idea how we should handle that for the internal tables that maintain state for pacos (and in the future for raft).

**Comment by kostja** [comment]
Right now we don't do paged reads from paxos. I believe we have once discussed that already and decided to handle on the monitor level: https://github.com/scylladb/scylla/issues/6329 https://github.com/scylladb/scylla-monitoring/issues/906

**Comment by vladzcloudius** [comment]
> AFAIK - internal queries are non prepared - we have that true for every where else in the system as well. 

It's probably everywhere else, except for Tracing and Audit/CDC (since it rides on the Tracing  framework), @slivne .

Please, see examples of the code in trace_keyspace_helper.cc and table_helper.hh.


> One of the issues is supporting the case in which they need to be re-prepaired.

This is a non-issue since it's been handled on the prepared_cache level long ago.
See examples in prepared_statements_cache.hh and the corresponding handling in table_helper.cc.
E.g. you can start here: https://github.com/scylladb/scylla/blob/master/table_helper.cc#L108

**Comment by vladzcloudius** [comment]
What's the status of this? 
We have a rather shameful Doc page: https://docs.scylladb.com/kb/increase-permission-cache/

How difficult is it to fix this? It looks pretty ridiculous when we have corresponding graphs on the "Advanced" Monitoring dashboard showing non-zero values simply because we send internal queries the way we teach our users NOT to do.

Non-zero values on these graphs should be red flags but those false alarms are making them useless.

@kostja You are right that we can hide this using corresponding labels in the most recent Monitoring + scylla versions. However this is even more ridiculous - using unprepared un-paged queries is always (!!!) a bad practice. Even for internal queries.

@slivne @harel-z @avikivity Can we, please, prioritize this? I believe this should take not more than 1-2 days to fix.

**Comment by mykaul** [comment]
@ptrsmrn - can we assess this? doesn't seem like a lot of work to implement.

**Comment by ptrsmrn** [comment]
Per @avikivity auth queries (which are internal queries) are prepared https://github.com/scylladb/scylla-enterprise/issues/1283#issuecomment-3101699299:
> All internal queries are prepared (the query text is used as the key to the cache).

So this task would only be limited to make auth internal queries paged, BUT first we'd need to check if that would bring any significant performance gain. 
@vladzcloudius is the issue about performance or metrics? Can you please attach some evidence (performance reports, which metrics are impacted).

**Comment by vladzcloudius** [comment]
> Per [@avikivity](https://github.com/avikivity) auth queries (which are internal queries) are prepared [scylladb/scylla-enterprise#1283 (comment)](https://github.com/scylladb/scylla-enterprise/issues/1283#issuecomment-3101699299):
> 
> > All internal queries are prepared (the query text is used as the key to the cache).
> 
> So this task would only be limited to make auth internal queries paged, BUT first we'd need to check if that would bring any significant performance gain. [@vladzcloudius](https://github.com/vladzcloudius) is the issue about performance or metrics? Can you please attach some evidence (performance reports, which metrics are impacted).

@ptrsmrn, please, try running cassandra-stress and observe corresponding metrics (`scylla_cql_unpaged_select_queries`) being bumped up.
To repeat what has already been mentioned on this GH issue 4 years ago: the fact that we are using unpaged reads internally while we teach our users to always use paged queries alone justifies fixing this.

I guess @denesb can give a few more reasons why using unpaged queries under any circumstances is a bad thing: one (which should be more than enough on its own) is that unpaged query memory consumption is not restricted unlike the one of the paged read.

**Comment by denesb** [comment]
Unpaged queries: bad.
They can be harmless for small data sizes and they are definitely simpler to use. But we should still migrate all our internal code to paged queries. Tables that are small today, can be big tomorrow, or at a certain customer.

**Comment by ptrsmrn** [comment]
Right, thanks. So do we have any customer issues where paged auth queries would actually help? And in general, do we have any data to motivate this change?

**Comment by vladzcloudius** [comment]
> Right, thanks. So do we have any customer issues where paged auth queries would actually help? And in general, do we have any data to motivate this change?

@ptrsmrn this is a ticking bomb and it needs to be fixed proactively. We don't really need any additional motivation for this fix beyond what's already been written above.

It's quite simple: scylla should **never** use unpaged queries internally.

@mykaul @avikivity, please, help getting this move forward since this discussion becomes cyclic.

**Comment by mykaul** [comment]
Ref. (not fixes) https://github.com/scylladb/scylladb/pull/25395

**Comment by ptrsmrn** [comment]
> Ref. (not fixes) [#25395](https://github.com/scylladb/scylladb/pull/25395)

It's just been merged, but doesn't solve the problem, there's many more unpaged queries: 39 matches for `execute_internal` (unpaged API) in `auth/`.
